frontend/components/ScatterChart.tsx

```diff
@@ -1,77 +1,142 @@
-import React, {FunctionComponent, useEffect} from 'react';
+import React, {FunctionComponent, useEffect, useMemo} from 'react';
 import {EChartOption} from 'echarts';
 import {WithStyled, primaryColor} from '~/utils/style';
 import useECharts from '~/hooks/useECharts';
 import {Dimension} from '~/types';
 
+const SYMBOL_SIZE = 12;
+
+type Point = {
+    name: string;
+    value: [number, number] | [number, number, number];
+};
+
 type ScatterChartProps = {
-    data?: ([number, number] | [number, number, number])[];
-    labels?: string[];
-    loading?: boolean;
-    dimension?: Dimension;
+    keyword: string;
+    loading: boolean;
+    points: Point[];
+    dimension: Dimension;
+};
+
+const assemble2D = (points: {highlighted: Point[]; others: Point[]}, label: EChartOption.SeriesBar['label']) => {
+    const createSeries = (name: string, data: Point[], patch?: {[k: string]: any}) => ({
```

Comment:
Maybe better type infer with `Partial<EChartOption['Series']>`, or use eslint-disabled to overleap type check 👀